### PR TITLE
JLine 3.24.1 and JAnsi 2.4.0.

### LINKED ISSUE #7371: Console broken (to varying degrees) for TERM=dumb
I am using sbt through the emacs `sbt-mode` (of which I'm the maintainer), which reports it self as a TERM=dumb. 

I was using sbt version 1.6.2 and noticed that the console is echoing erroneously and came to write a bug report on that. While doing so, I decided to upgrade to latest sbt (1.9.4) so as not to report a bug against an old version. To do so, I created a new clean project and there 1.6.2 as well as 1.9.4 are even more broken, so much so that at the moment the console is unusable in term=DUMP in 1.9.4.

I will thus report both behaviors (partially broken 1.6.2 and completely broken 1.9.4) here.

The partial echo problem was also reproduced on 1.8.3 by @rossabaker (in the `Cats` project)

## steps

1. Create an empty new project folder and move  to it
2. run `env TERM=dumb sbt consoleQuick`
3. type: `"foo"`
4. you will see the following:

```
scala> "foo"
foo"
<console>:1: error: unclosed string literal
       foo"
```
5. type `vval x = 1` [sic!]

```
scala> vval x = 1
val x = 1
x: Int = 1
```

If at step 2. you run  `env TERM=dumb sbt` and then in sbt-shell type `consoleQuick` it works differently. In this case, there is full echo, but the first character of the input is not skipped. The scala version in the clean project is the default `2.12.15`

For 1.6.2 in my project (with scala version `2.13.8`), I get the following

1. run `env TERM=dumb sbt consoleQuick`
2. type `"foo"` and there is no echo
```
scala> "foo"
val res1: String = foo
```
3. type `"foo)bar"` and there is echo until `)`
```
scala> "foo)bar"
"foo)val res2: String = foo()
```

4. type `"foo)bar(}zot{"`
```
scala> "foo)bar(}zot{"
"foo)bar(}val res11: String = foo)bar(}zot{
```

In the latter case, the results do not depend on whether the console was started from the command line or from sbt shell.

## problem

When in the DUMB terminal mode, there should be no echo.  However...

1.9.4 in clean and empty project folder: there is full echo (the line right after the input was given) and the console is skipping the first character of the input.

1.6.2 in my project: there is no echo on most input, but if the input contains any closing parenthesis or brackets (`)}]`), the terminal echoes the part of the expression until the last closing bracket. Note that the closing parenthesis can be even inside a string

## expectation

There should be no echo in any case and the terminal should not skip the first character of input.

### PR DESCRIPTION
**Pull request by hvesalai**
Fix #7371

- https://github.com/fusesource/jansi/releases/tag/jansi-2.4.1
- https://github.com/jline/jline3/releases/tag/jline-parent-3.24.1

### CODE REVIEW

**Comment by hvesalai** [comment]
This PR is still waiting for an update on sbt/jline2.

File: project/Dependencies.scala [hunk 1]
```diff
@@ -87,13 +87,13 @@ object Dependencies {
   // JLine 3 version must be coordinated together with JAnsi version
   // and the JLine 2 fork version, which uses the same JAnsi
   val jline = "org.scala-sbt.jline" % "jline" % "2.14.7-sbt-a1b0ffbb8f64bb820f4f84a0c07a0c0964507493"
-  val jline3Version = "3.19.0"
+  val jline3Version = "3.24.0"
```
Comment (xuwei-k) [inline project/Dependencies.scala]: 🤔 

- https://github.com/jline/jline3/releases/tag/jline-parent-3.24.0
- https://github.com/jline/jline3/commit/a2ab0474a023e2ad062978177fc6f5b3c8e69501

**Comment by SethTisue** [comment]
JLine 3.24 doesn't work on JDK 8 — better stop at 3.23 (reference: https://github.com/scala/scala-dev/issues/856)

(I think that's what Yoshida-san is trying to alert us to, above)

**Comment by hvesalai** [comment]
JLine 3.24 contains the fix I need. I could ask the developer of JLine to backport it to 3.23.1, but I don't know how he will feel about it since no other such patch releases have been done in the past. I made the request: https://github.com/jline/jline3/issues/888

Optionally we could create our own branch of JLine3 as we have done with JLine2, but at some point cleaning up the whole JLine dependency thing should be planned. I can help if there is a clear direction we are going to take.

Does this imply that scala will try to support JDK8 for ever?

**Comment by SethTisue** [comment]
> Does this imply that scala will try to support JDK8 for ever?

Unclear. So far, keeping support hasn't been painful. Perhaps at some point we'll decide a tipping point has been reached for the costs of continuing support have become too high.

Anyway, it's a moot point for now, as I see that the JLine 3 maintainer just said that they will restore JDK 8 support in 3.24.1. So let's wait for that.

**Comment by hvesalai** [comment]
Updated the PR with JLine 3.24.1 (was 3.24.0)

**Comment by hvesalai** [comment]
@eed3si9n did you get a chance to release sbt/jline2 already? If you could tell the version number, I can update it to this PR.

**Comment by eed3si9n** [comment]
Yea, I released it - https://repo1.maven.org/maven2/org/scala-sbt/jline/jline/2.14.7-sbt-9c3b6aca11c57e339441442bbf58e550cdfecb79/

**Review by eed3si9n — APPROVED** [review]
Thanks!

**Comment by lrytz** [comment]
When I take current 1.10.x, run `sbt publishLocal`, then `rm -rf ~/.sbt/boot/scala-2.12.18/org.scala-sbt/sbt/1.9.7-SNAPSHOT` and then try `1.9.7-SNAPSHOT` in a test project, I get

```
➜ sbtproj sbt
[info] [launcher] getting org.scala-sbt sbt 1.9.7-SNAPSHOT  (this may take some time)...
java.lang.NoClassDefFoundError: org/jline/terminal/spi/JansiSupport
	at java.base/java.lang.ClassLoader.defineClass1(Native Method)
	at java.base/java.lang.ClassLoader.defineClass(ClassLoader.java:1027)
	at java.base/java.security.SecureClassLoader.defineClass(SecureClassLoader.java:150)
	at java.base/java.net.URLClassLoader.defineClass(URLClassLoader.java:524)
	at java.base/java.net.URLClassLoader$1.run(URLClassLoader.java:427)
	at java.base/java.net.URLClassLoader$1.run(URLClassLoader.java:421)
	at java.base/java.security.AccessController.doPrivileged(AccessController.java:714)
	at java.base/java.net.URLClassLoader.findClass(URLClassLoader.java:420)
	at java.base/java.lang.ClassLoader.loadClass(ClassLoader.java:593)
	at java.base/java.lang.ClassLoader.loadClass(ClassLoader.java:580)
	at java.base/java.lang.ClassLoader.loadClass(ClassLoader.java:580)
	at java.base/java.lang.ClassLoader.loadClass(ClassLoader.java:580)
	at java.base/java.lang.ClassLoader.loadClass(ClassLoader.java:526)
	at sbt.internal.util.JLine3$.<init>(JLine3.scala:53)
	at sbt.internal.util.JLine3$.<clinit>(JLine3.scala)
	at sbt.internal.util.Terminal$.newConsoleTerminal(Terminal.scala:805)
	at sbt.internal.util.Terminal$.withStreams(Terminal.scala:386)
	at sbt.xMain$.withStreams$1(Main.scala:88)
	at sbt.xMain$.run(Main.scala:123)
	at java.base/jdk.internal.reflect.DirectMethodHandleAccessor.invoke(DirectMethodHandleAccessor.java:103)
	at java.base/java.lang.reflect.Method.invoke(Method.java:580)
	at sbt.internal.XMainConfiguration.run(XMainConfiguration.java:59)
	at sbt.xMain.run(Main.scala:47)
	at xsbt.boot.Launch$.$anonfun$run$1(Launch.scala:149)
	at xsbt.boot.Launch$.withContextLoader(Launch.scala:176)
	at xsbt.boot.Launch$.run(Launch.scala:149)
	at xsbt.boot.Launch$.$anonfun$apply$1(Launch.scala:44)
	at xsbt.boot.Launch$.launch(Launch.scala:159)
	at xsbt.boot.Launch$.apply(Launch.scala:44)
	at xsbt.boot.Launch$.apply(Launch.scala:21)
	at xsbt.boot.Boot$.runImpl(Boot.scala:78)
	at xsbt.boot.Boot$.run(Boot.scala:73)
	at xsbt.boot.Boot$.main(Boot.scala:21)
	at xsbt.boot.Boot.main(Boot.scala)
Caused by: java.lang.ClassNotFoundException: org.jline.terminal.spi.JansiSupport
	at java.base/java.net.URLClassLoader.findClass(URLClassLoader.java:445)
	at java.base/java.lang.ClassLoader.loadClass(ClassLoader.java:593)
	at java.base/java.lang.ClassLoader.loadClass(ClassLoader.java:526)
	... 34 more
[error] [launcher] error during sbt launcher: java.lang.NoClassDefFoundError: org/jline/terminal/spi/JansiSupport
```

@hvesalai  / @eed3si9n  does it work for you?

**Comment by lrytz** [comment]
macOS 14.2, `OpenJDK Runtime Environment Temurin-17.0.9+9 (build 17.0.9+9)`

**Comment by eed3si9n** [comment]
I can reproduce the problem. The fix is to bump the version number to `1.10.0-SNAPSHOT`. This is likely side effect of multi-repo (Zinc etc) using util packages from sbt/sbt but with 1.10.0-M1.

**Comment by lrytz** [comment]
Thanks Eugene, that fixes it 👍

**Pull request merged**